src/NuGet.Clients/NuGet.VisualStudio.Common/ServiceProviderExtensions.cs

```diff
@@ -25,5 +25,13 @@ public static class ServiceProviderExtensions
         {
             return site.GetServiceAsync<SComponentModel, IComponentModel>();
         }
+
+        public static async Task<TInterface> GetFreeThreadedServiceAsync<TService, TInterface>(this IAsyncServiceProvider site) where TInterface : class
+        {
+#pragma warning disable RS0030 // Do not used banned APIs
+            return (TInterface)await site.GetServiceAsync(typeof(TService));
```

Comment:
Why not use `GetServiceAsync<TService, TInterface>()`?

Reply:
That extension method switches to the UI thread when it casts to TInterface, and for free threaded service acquisition, we don't want that.

Reply:
I think that should be documented in a comment, because there's no way most people would realize this, and unlikely I'll remember in 3 months.

Is there some way to determine when a service is free threaded and therefore safe to use on this method?  COM casting happens implicitly, and the only time you know you've done it wrong is when the UI thread is blocked on the background thread trying to do the cast, and there's a deadlock.

Reply:
> I think that should be documented in a comment, because there's no way most people would realize this, and unlikely I'll remember in 3 months.

I can document it. 

> Is there some way to determine when a service is free threaded and therefore safe to use on this method?

No, not really. 
I have asked a lot of questions about it in the VS Threading group (see: VS services and free threadedness) but there's a very short list of confirmed ones: 

```cs
SProfferService

SLocalRegistry

SUIHostLocale

SVsActivityLog

SVsResourceManager

SVsSettingsManager

SVsInvalidateCachedCommandState

SVsUIContextMonitor

SAsyncServiceProvider

SProfferAsyncService

SVsTaskSchedulerService

```

SComponentModel and SVsBrokeredServiceContainer can be added to that list as well.

Reply:
Added a comment. 

Resolving this thread. Feel free to comment again if you don't agree.